### DISCUSSION TITLE
『浅入浅出』MySQL 和 InnoDB · /mysql-innodb

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#254] [Announcements]
https://draveness.me/mysql-innodb

**Comment by baoqp** [comment]
文章不仅内容好，排版、图片都很好看。

**Comment by leizhiyuan** [comment]
想问一下,作者的图都是用什么画的?非常漂亮

**Comment by draveness** [comment]
你看看文章最后的说明

**Comment by leizhiyuan** [comment]
> @Draveness
> 你看看文章最后的说明
> 
> On Dec 1, 2017, 2:07 PM +0800, Lei Zhiyuan <notifications@github.com>, wrote:
> > 想问一下,作者的图都是用什么画的?非常漂亮
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> 

下意识以为这里是转载和授权..谢谢~

**Comment by draveness** [comment]
我之后加个更明显的说明吧🌚

**Comment by ximeng123** [comment]
支持一波，内容可以的

**Comment by luodonger** [comment]
非常感谢，写得真好！！！

**Comment by StevenWash** [comment]
非常赞

**Comment by lancer1981** [comment]
说的非常清楚

**Comment by walterzhaoJR** [comment]
感谢博主分享，写的很棒。有个地方确认想和你确认下：锁的算法 中的 Record Lock 中  如果我们使用 id 或者 last_name 作为 SQL 中 WHERE 语句的过滤条件，那么 InnoDB 就可以通过索引建立的 B+ 树找到行记录并添加索引（锁）

**Comment by wlkevin** [comment]
想请教个问题，对于varchar与BLOB类型的列采用指针方式实现存储，但是对于如果最后一行数据超出了页剩余空间大小，该如何处理呢？

**Comment by draveness** [comment]
> @wlkevin
> 想请教个问题，对于varchar与BLOB类型的列采用指针方式实现存储，但是对于如果最后一行数据超出了页剩余空间大小，该如何处理呢？

开一个新的页啊。。

**Comment by MA-Aaron** [comment]
您写的太好了，图文并茂

**Comment by throw-panic** [comment]
前辈您好，您写的文章很详细。本人在校生，想转到自己的博客学习之用，会注明来源出处，可以吗？

**Comment by draveness** [comment]
> @MAIRONGRUI
> 前辈您好，您写的文章很详细。本人在校生，想转到自己的博客学习之用，会注明来源出处，可以吗？

> 本作品采用知识共享署名 4.0 国际许可协议进行许可。 转载时请注明原文链接，图片在使用时请保留图片中的全部内容，可适当缩放并在引用处附上图片所在的文章链接，图片使用 Sketch 进行绘制。

**Comment by rainyInSeptember** [comment]
2-200行，innodb 那本书上写的7992行吧？！

**Comment by alvin-mei** [comment]
不可重复读是指同一事务中两次读取同一数据得到的内容不一致，因此能够使用行锁来解决。

**Comment by draveness** [comment]
> @alvin-mei
> 不可重复读是指同一事务中两次读取同一数据得到的内容不一致，因此能够使用行锁来解决。

查了下，发现这里写的确实有问题，不可重复读指的是对同一数据的操作，所以行锁确实可以解决，我之后会改下文章的内容，多谢

https://en.wikipedia.org/wiki/Isolation_%28database_systems%29#Non-repeatable_reads

**Comment by holyliao** [comment]
写得真好

**Comment by Hhhhhui** [comment]
很NB的文章

**Comment by zmknshn** [comment]
太感谢了

**Comment by zhyzhyzhy** [comment]
支持一下
不过有个问题，READ COMMITTED的隔离级别 
> 只对记录加记录锁，而不会在记录之间加间隙锁

但是这个是不是得区分下是快照读还是当前读，我理解只有当前读才会加锁

文档 [https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html](https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html)

**Comment by draveness** [comment]
> @zhyzhyzhy
> 支持一下
> 不过有个问题，READ COMMITTED的隔离级别 
> > 只对记录加记录锁，而不会在记录之间加间隙锁
> 
> 但是这个是不是得区分下是快照读还是当前读，我理解只有当前读才会加锁
> 
> 文档 [https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html](https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html)

是的，但是这里说的其实是在事务中的『当前读』

**Comment by 290932857** [comment]
你好，首先支持一下，写的都比较浅显易懂，其次有个问题不太明白，想问下，说到索引的时候，每次索引都查询到数据行所在的page，然后page里面循环定位到具体数据，但是后面聚集索引讲到在每个叶子节点处存储当前键对应行数据的所有数据，这个时候应该就不是定位到page了吧？ 另外是不是这个时候数据就存储了两份？聚集索引存了一份，page存储了一份？望解答，谢谢。

**Comment by draveness** [comment]
> @290932857
> 你好，首先支持一下，写的都比较浅显易懂，其次有个问题不太明白，想问下，说到索引的时候，每次索引都查询到数据行所在的page，然后page里面循环定位到具体数据，但是后面聚集索引讲到在每个叶子节点处存储当前键对应行数据的所有数据，这个时候应该就不是定位到page了吧？ 另外是不是这个时候数据就存储了两份？聚集索引存了一份，page存储了一份？望解答，谢谢。

能不能别发这么多次啊，索引本来就是按照页去存储的，聚集索引你可以理解成一堆页组成的，只存了一份

**Comment by wuleizi** [comment]
非常赞

**Comment by FFFjunrui** [comment]
不可重复读是因为没有加行锁，可能会出现第二次读取被其他事物修改，同样会出现幻读
对该行加锁后隔离级别变更为可重复读，同样也会出现幻读，使用MVCC解决幻读的问题

**Comment by draveness** [comment]
> @FFFjunrui
> 不可重复读是因为没有加行锁，可能会出现第二次读取被其他事物修改，同样会出现幻读
> 对该行加锁后隔离级别变更为可重复读，同样也会出现幻读，使用MVCC解决幻读的问题

你是在总结么..还是在提问题，MySQL 通过间隙锁（Next-Key）解决幻读的问题

**Comment by ghost** [comment]
讲不可重复读和read-commited隔离级别的时候前后有些矛盾，前面总结的时候是说没有加间隙锁，后面举的例子是记录锁（id=3。update的语句和查询的语句相同，不在间隙中）；

只对记录加记录锁，而不会在记录之间加间隙锁，所以允许新的记录插入到被锁定记录的附近，所以再多次使用查询语句时，可能得到不同的结果（Non-Repeatable Read）；

**Comment by draveness** [comment]
> 讲不可重复读和read-commited隔离级别的时候前后有些矛盾，前面总结的时候是说没有加间隙锁，后面举的例子是记录锁（id=3。update的语句和查询的语句相同，不在间隙中）；

能展开说一下问题么，我对照了半天原文和你在这里提的问题...

**Comment by DelbertTse** [comment]
你好，先膜拜一下大佬，然后想请教一个问题，文章中说的B+Tree索引的比较次数是树的高度，但是我理解的是树的高度是IO的次数，并不是实际比较的次数，使用B+Tree的原因也是因为IO次数稳定，且少于B树，那比较次数的话应该怎么计算呢。

**Comment by draveness** [comment]
> 我理解的是树的高度是IO的次数

在大多数情况下，索引的非叶节点基本都会缓存在内存中，也就是不需要随机 IO，不过最差的情况就是 IO 次数和树高相同

> 并不是实际比较的次数

那你认为比较次数和什么有关呢

**Comment by vuuihc** [comment]
『READ COMMITED：只对记录加记录锁，而不会在记录之间加间隙锁』，这里跟下面有些矛盾？下面说READ COMMITED不会加行锁。

**Comment by draveness** [comment]
> @vuuihc
> 『READ COMMITED：只对记录加记录锁，而不会在记录之间加间隙锁』，这里跟下面有些矛盾？下面说READ COMMITED不会加行锁。

在哪里写的不会加行锁？我没找到..

**Comment by bspp1314** [comment]
在Mysql8.0中已经不再使用frm文件

**Comment by SimpleIto** [comment]
文章写得很好，深入浅出的方式 以及知识的深度也差不多是我自己追求的理想的学习水平。最近临近毕业有些迷茫，可以请问下像博主这样的人在哪里工作呢？

**Comment by mrdear** [comment]
赞一个

**Comment by draveness** [comment]
> @亦非台 请教一个问题，每个表mysql都会创建一个聚簇索引的B+树结构，数据会存放在叶子节点上，那么如果查询语句中不能用到索引，就不会走聚簇索引的B+树，那么从哪里查询数据呢？

这个问题是删了么....

**Comment by a601942905git** [comment]
> @draveness
> > @亦非台 请教一个问题，每个表mysql都会创建一个聚簇索引的B+树结构，数据会存放在叶子节点上，那么如果查询语句中不能用到索引，就不会走聚簇索引的B+树，那么从哪里查询数据呢？
> 
> 这个问题是删了么....

是的，后来想了下。聚簇索引是mysql存放数据的方式，数据存放在聚簇索引对应B+树的叶子节点上。一开始的疑惑是创建聚簇索引会将数据存放在叶子节点上，如果不创建聚簇索引，mysql也会存放数据，那么创建聚簇索引，就会存在存储2份一样的数据，后来了解到mysql一定会创建一个聚簇索引，数据只会存放在聚簇索引的叶子节点上，就消除了该疑问。

**Comment by sowater** [comment]
请教一下，作者是用什么画图的，感觉挺不错，求推荐

**Comment by draveness** [comment]
> @sowater  请教一下，作者是用什么画图的，感觉挺不错，求推荐

你看看文章结尾还有上面的评论？

**Comment by hhyvs111** [comment]
RR级别下解决幻读不止是用next lock解决的吧？
- 快照读（select）是用mvcc版本控制解决的；
- 当前读（select for update）才是用next lock解决的。

**Comment by draveness** [comment]
> RR级别下解决幻读不止是用next lock解决的吧？
> 
> * 快照读（select）是用mvcc版本控制解决的；
> * 当前读（select for update）才是用next lock解决的。

你说的 MVCC 和隔离级别里的幻读没关系吧...

**Comment by ansionfor** [comment]
“就可以通过索引建立的 B+ 树找到行记录并添加索引”。索引应该是锁吧

**Comment by draveness** [comment]
> @ansionfor
> “就可以通过索引建立的 B+ 树找到行记录并添加索引”。索引应该是锁吧

对，我改一下

**Comment by bigrotor187** [comment]
为啥我看官方文档看不出啥重点，主要是看Reference Manual吗？

**Comment by draveness** [comment]
> 为啥我看官方文档看不出啥重点，主要是看Reference Manual吗？

还有几本 MySQL 相关的书籍

**Comment by ghost** [comment]
> > RR级别下解决幻读不止是用next lock解决的吧？
> > 
> > * 快照读（select）是用mvcc版本控制解决的；
> > * 当前读（select for update）才是用next lock解决的。
> 
> 你说的 MVCC 和隔离级别里的幻读没关系吧...

有关系的。具体可以看innodb documents里讲隔离级别和两种reading。lock相关的是locking-read，mvcc来保证none-locking read

**Comment by vc60er** [comment]
原文：
> RAED UNCOMMITED：使用查询语句不会加锁，可能会读到未提交的行（Dirty Read）；  
> READ COMMITED：只对记录加记录锁，而不会在记录之间加间隙锁，所以允许新的记录插入到被锁定记录的附近，所以再多次使用查询语句时，可能得到不同的结果（Non-Repeatable Read）；

脏读和不可重复读的原因好像有问题. 我理解是：

RAED UNCOMMITED：事务没有commit就落地盘，可能造成其他事务会读到未提交的行（Dirty Read）；

READ COMMITED：事务中select没有行锁，可能会读取到其他事务commit的修改。造成不可重复读（Non-Repeatable Read）；

并且，与下文中描述的原因也不一致

原文：
> 脏读.  
> 在一个事务中，读取了其他事务未提交的数据。   
> 不可重复读.  
> 在一个事务中，同一行记录被访问了两次却得到了不同的结果。

**Comment by OrdinaryYZH** [comment]
"ISO 和 ANIS SQL 标准制定了四种事务隔离级别"这里的ANIS拼写错了 :)

---

*2020-06-06 UPDATES: 已修复*

**Comment by OrdinaryYZH** [comment]
关于文中幻读中的例子，事务1中再查询应该可以查询出事务2提交的结果吧，因为事务1的查询范围不单止1个，不属于REPEATABLE READ了。我是参考这篇文章中举例的：https://lotabout.me/2020/QQA-Isolation-Level-of-Database#p3-phantom

**Comment by wmyskxz** [comment]
大大 图挂了

**Comment by draveness** [comment]
> @wmyskxz
> 大大 图挂了

我这里没有

**Comment by darkfireworld** [comment]
建议补充一下事务隔离级别（RC/RR）的InnoDB中MVCC机制。
快照读：实现依赖于MVCC（事务版本号，ReadView生成时机，UNDO LOG等），推荐：https://zhuanlan.zhihu.com/p/64576887
当前读：使用lock机制（具体依赖索引和主键索引加锁等），推荐：https://www.jianshu.com/p/346d3c30eb2c

**Comment by nudt681** [comment]
你好，请教一下，我现在看数据页清楚了，就是叶子节点的存储方式，但是非叶子节点的存储是什么结构呢，是否和叶子节点一样

**Comment by ymyqwe** [comment]
写的真是浅显易懂，赞一个

**Comment by Coolxc** [comment]
图挂了。。。

**Comment by EeeMt** [comment]
写得很棒

**Comment by GabyCheng** [comment]
写的真好！

**Comment by bigrotor187** [comment]
直接Google发现排第一的是直接赤裸裸抄袭的 http://blog.itpub.net/69917606/viewspace-2657574/

**Comment by draveness** [comment]
> 直接Google发现排第一的是直接赤裸裸抄袭的 http://blog.itpub.net/69917606/viewspace-2657574/

谢谢你的提醒，不过这种真的很难办....

**Comment by MingdaLin** [comment]
你好，next-key锁是如何解决幻读问题的呢？可以再补充一下吗？十分感谢🙏

**Comment by draveness** [comment]
> 你好，next-key锁是如何解决幻读问题的呢？可以再补充一下吗？十分感谢🙏

哪里不明白啊

**Comment by funye** [comment]
学习了，不过文章这句 “REPEATABLE READ 和 READ UNCOMMITED 其实是矛盾的”话是不是写错了， 应该是REPEATABLE READ 和 READ COMMITED 是矛盾的吧 ？

**Comment by draveness** [comment]
> 学习了，不过文章这句 “REPEATABLE READ 和 READ UNCOMMITED 其实是矛盾的”话是不是写错了， 应该是REPEATABLE READ 和 READ COMMITED 是矛盾的吧 ？

为什么是矛盾的，可以展开说一下么，这篇文章有点久了

**Comment by wolf-911** [comment]
部分图片不显示呢

**Comment by draveness** [comment]
> @wolf-911
> 部分图片不显示呢

没问题啊

**Comment by wolf-911** [comment]
截图已发到您公众号，不知道您是否能看到

**Comment by ydzh** [comment]
不可重复读的原因就是，在 READ COMMITED 的隔离级别下，存储引擎不会在查询记录时添加行锁，锁定 id = 3 这条记录。
上面的是语句没有加锁导致的吧。主是因为read commited 没有记录第一次读的事务id，每次都取最新事务

**Comment by gongqij** [comment]
"不可重复读的原因就是，在 READ COMMITED 的隔离级别下，存储引擎不会在查询记录时添加行锁，锁定 id = 3 这条记录。"这句话描述好像不太准确，select 。。。from table这种查询语句是一致性非锁定读操作，不会添加行锁，而不可重复读的原因是READ COMMITED隔离级别下读取的是最新的行数据版本，REPEATABLE READ隔离级别下读取的是事务开始的行数据版本。还是我理解错了，请指教。。

**Comment by asddongmen** [comment]
```
READ COMMITED：只对记录加记录锁，而不会在记录之间加间隙锁，所以允许新的记录插入到被锁定记录的附近，所以再多次使用查询语句时，可能得到不同的结果（Non-Repeatable Read）；
```
感觉这里应该修正一下，要不然会引起误解。
1.在 READ COMMITED 隔离级别下，单纯的读操作并不会加行级锁，并且在同一个事务内每次读操作都会读当前最新快照，因此可能会出现不可重复读。
2.在 READ COMMITED 隔离级别下，`SELECT with FOR UPDATE or FOR SHARE` 才会加行级锁；但是由于没有 `gap lock` 和 `next-key lock` 会产生幻读。

refer to : [Transaction Isolation Levels](https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html)

**Comment by ThreeBearsDan** [comment]
```
READ COMMITED：只对记录加记录锁，而不会在记录之间加间隙锁，所以允许新的记录插入到被锁定记录的附近，所以再多次使用查询语句时，可能得到不同的结果（Non-Repeatable Read）；
```
这段对不可重复读的原因阐述是错误的吧？写成了幻读的原因

**Comment by chenghung** [comment]
淺顯易懂
這篇文章寫的非常棒

**Comment by howoii** [comment]
大佬图挂了，貌似有些图不支持https

**Comment by draveness** [comment]
> @howoii
> 大佬图挂了，貌似有些图不支持https

我这里看是没有问题的

**Comment by howoii** [comment]
> @draveness
> > @howoii
> > 大佬图挂了，貌似有些图不支持https
> 
> 我这里看是没有问题的

我这边的报错：GET https://raw.githubusercontent.com/Draveness/Analyze/master/contents/Database/images/mysql/mysql.png net::ERR_CERT_INVALID
